### PR TITLE
cleanup: set targeted dom-change listener

### PR DESCRIPTION
**Pull request by stephanwlee**
Instead of component-level listener for dom-change, we now put the
listener on the specific template (dom-if or dom-repeat) that we care
about.

### CODE REVIEW

File: tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html [hunk 1]
```diff
@@ -719,12 +722,14 @@
         );
       },
       _matchListEntryColorUpdated(event) {
```
Comment (psybuzz) [inline tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html]: We can now drop `event` param

File: tensorboard/components/tf_dashboard_common/tf-filterable-checkbox-list.html [hunk 2]
```diff
@@ -58,7 +58,11 @@
           No matches
         </div>
       </template>
-      <template is="dom-repeat" items="[[_itemsMatchingRegex]]">
+      <template
+        is="dom-repeat"
+        items="[[_itemsMatchingRegex]]"
+        on-dom-change="_synchronizeColors"
```
Comment (psybuzz) [inline tensorboard/components/tf_dashboard_common/tf-filterable-checkbox-list.html]: Is this deep / Does this trigger when users check/uncheck a checkbox?
If so, maybe consider an observer: `_synchronizeColors(_itemsMatchingRegex)`

File: tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html [hunk 3]
```diff
@@ -719,12 +722,14 @@
...
-        const [maybeDomRepeat] = event.path;
-        if (maybeDomRepeat.nodeName !== 'DOM-REPEAT') return;
+        const domRepeat = this.$$('#match-list-repeat');
```
Comment (psybuzz) [inline tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html]: Curious why the change from `event.path` to this selector.  Is .path, .target not reliable?

Comment (stephanwlee) [reply tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html]: We only needed `event.path` because `dom-if` also emits the same event and we wanted to discern the events. Now that we have more localized listener, we simply don't need one.

File: tensorboard/components/tf_dashboard_common/tf-filterable-checkbox-list.html (hunk 2 above)
Comment (stephanwlee) [reply tensorboard/components/tf_dashboard_common/tf-filterable-checkbox-list.html]: I do not follow. Why is that required?

Comment (psybuzz) [reply tensorboard/components/tf_dashboard_common/tf-filterable-checkbox-list.html]: It seems it may be unnecessary to call `_synchronizeColors` on every single dom change.  It depends on the items, but not on whether a checkbox is checked.  If a user checks a checkbox and triggers dom to change, then we don't actually need to call it.

File: tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html (hunk 3 above)
Comment (psybuzz) [reply tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html]: The `event.path` works before and after, just curious whether we gain something by switching to a querySelector.  Either way is fine with me, though

File: tensorboard/components/tf_dashboard_common/tf-filterable-checkbox-list.html (hunk 2 above)
Comment (stephanwlee) [reply tensorboard/components/tf_dashboard_common/tf-filterable-checkbox-list.html]: > unnecessary to call _synchronizeColors on every single dom change.

this is the part I am not getting. `dom-change` is a feature of dom-repeat and dom-if. Are you saying paper-checkbox uses that and thus will fire bunch of events?

Comment (psybuzz) [reply tensorboard/components/tf_dashboard_common/tf-filterable-checkbox-list.html]: > dom-change is a special event of dom-repeat and dom-if

Ahh, things make sense then, I was wrong!
For my own understanding: so before, when a component (not dom-repeat/if) has a `listeners: {'dom-change': ...}`, that means it is listening specifically for events dispatched by dom-repeat/if.
